### PR TITLE
Farms near freshwater

### PR DESCRIPTION
**Pull request by HadeanLake**
2020-08-31-20:23:02_1920x1080_scrot

91900151-e9d50700-eca6-11ea-951c-b077425ab1d3

* Farm now has unique "Can also be built on tiles adjacent to fresh water"
https://civilization.fandom.com/wiki/Farm_(Civ5)
Even snow: https://www.reddit.com/r/civ/comments/11qtko/farms_on_snow/
Hill+Farm, Snow+Farm, Tundra+Farm sprites by THE BUCKETEER and Ravignir from discord
* fields improvingTech and improvingTechStats converted into uniques
e.g.:
"[+1 Production] once [Scientific Theory] is discovered"
"[+1 Food] on [fresh water] tiles once [Civil Service] is discovered"
"[+1 Food] on [non-fresh water] tiles once [Fertilizer] is discovered"
* "Provides a one-time Production bonus to the closest city when cut down" unique for forest
Provides 20 if forest is cut next to city, 20 if one tile away, 15 if two tiles away, then 10, 5, 0
2/3 of that if tile is not owned by worker's civ - 13, 13, 10, 6, 3, 0

### CODE REVIEW

File: core/src/com/unciv/logic/map/TileInfo.kt [hunk 1]
```diff
@@ -243,9 +243,15 @@ open class TileInfo {
         if (hasViewableResource(observingCiv) && getTileResource().improvement == improvement.name)
             stats.add(getTileResource().improvementStats!!.clone()) // resource-specific improvement
 
+        // As of 3.10.5 This is to be deprecated and converted to "[stats] once [tech] is discovered" - keeping it here to that mods with this can still work for now
         if (improvement.improvingTech != null && observingCiv.tech.isResearched(improvement.improvingTech!!))
             stats.add(improvement.improvingTechStats!!) // eg Chemistry for mines
 
+        for (unique in improvement.uniqueObjects) when (unique.placeholderText) {
+            "[] once [] is discovered" -> if (observingCiv.tech.isResearched(unique.params[1])) stats.add(Stats.parse(unique.params[0]))
+            "[] on fresh water tiles once [] is discovered" -> if (observingCiv.tech.isResearched(unique.params[1]) && isAdjacentToFreshwater) stats.add(Stats.parse(unique.params[0]))
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/TileInfo.kt:252]: "[] on [] tiles once [] is discovered"
then "fresh water" and "non-fresh water tiles" are parameters, and "[] tiles" can be united with the other tile parameter check function

Comment (HadeanLake) [reply core/src/com/unciv/logic/map/TileInfo.kt]: unified

File: core/src/com/unciv/logic/map/MapUnit.kt [hunk 2]
```diff
@@ -399,6 +404,16 @@ class MapUnit {
         tile.improvementInProgress = null
     }
 
+    private fun tryProvideProductionToClosestCity()
+    {
+        val tile = getTile()
+        val closestCity = civInfo.cities.minBy { it.location.dst(tile.position) }
+        if (closestCity == null) return
+        val productionToAdd = (25 - closestCity.location.dst(tile.position) * 4).toInt()
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/MapUnit.kt:412]: Where does this come from? Couldn't find any sources about this.
But if is is dependent on distance to city, we should use closestCity.getCenterTile().aerialDistanceTo(tile)

File: core/src/com/unciv/logic/map/MapUnit.kt [hunk 3]
```diff
@@ -399,6 +404,16 @@ class MapUnit {
...
+        if (productionToAdd > 0) closestCity.cityConstructions.addProductionPoints(productionToAdd)
+        //civInfo.addNotification("[$productionToAdd] Production points added towards current construction in [${closestCity.name}]", closestCity.location, Color.YELLOW)
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/MapUnit.kt:414]: And yes, we should enable this notification :)

File: core/src/com/unciv/models/ruleset/tech/Technology.kt [hunk 4]
```diff
@@ -24,6 +24,20 @@ class Technology {
...
+        for (improvement in ruleset.tileImprovements.values) for ( unique in improvement.uniqueObjects
+                .filter { it.placeholderText in setOf("[] once [] is discovered", "[] on [] tiles once [] is discovered")
+                        && it.params.last() == name }) {
+            val key = if (unique.params.size == 2 ) unique.params[0] else "[${unique.params[0]}] on [${unique.params[1]}] tiles"
+            if (!mapOfImprovedImprovements.containsKey(key)) mapOfImprovedImprovements[key] = ArrayList()
+            mapOfImprovedImprovements[key]!!.add(improvement.name)
+        }
+        for ( improvements in mapOfImprovedImprovements) {
+            val impimpString = improvements.value.joinToString(", ") { it.tr() } +
+                    " {provide" + (if (improvements.value.size == 1) "s" else "") + "} " + improvements.key.tr()
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/tech/Technology.kt:37]: I really, really don't like this method of pluralization... there must be a way to have another sentence that doesn't require this.

Comment (yairm210) [reply core/src/com/unciv/models/ruleset/tech/Technology.kt:37]: And yes I'm aware that this was probably copied from my code (impimpString rings a bell) but I was wrong, this is yukky

File: core/src/com/unciv/logic/map/MapUnit.kt (hunk 3 above)
Comment (HadeanLake) [reply core/src/com/unciv/logic/map/MapUnit.kt]: enabled

File: core/src/com/unciv/logic/map/MapUnit.kt (hunk 2 above)
Comment (HadeanLake) [reply core/src/com/unciv/logic/map/MapUnit.kt]: Yes, dependent on distance.
jungles yield 20 in this mod https://steamcommunity.com/sharedfiles/filedetails/?id=103271218 - same as forest author stated in description

https://steamcommunity.com/app/8930/discussions/0/46476145384069876/
>2 tiles away is still 20 and on third one is 17
https://forums.civfanatics.com/threads/how-far-from-city-can-chop-forests-for-production.487398/
> Haven't tested it, but I think I've chopped forests more than 5 tiles away (but maybe not). I've definitely chopped outside my culture borders and that gives fewer hammers for the same distance from the city center.

Current guess is` if (distance < 2) 20 else (20 - (distance - 2) * 3).toInt()`
probably some modifier for not being in your borders. Maybe it starts to decrease outside

File: core/src/com/unciv/models/ruleset/tech/Technology.kt (hunk 4 above)
Comment (HadeanLake) [reply core/src/com/unciv/models/ruleset/tech/Technology.kt:37]: There is this line:
2020-09-01-22:54:50_1920x1080_scrot

So maybe
Stats [on tiles] from improvements: [list]
2020-09-01-22:55:03_1920x1080_scrot

Comment (yairm210) [reply core/src/com/unciv/models/ruleset/tech/Technology.kt:37]: Looks OK to me

Comment (HadeanLake) [reply core/src/com/unciv/models/ruleset/tech/Technology.kt:37]: Conflicting files
android/assets/game.atlas
android/assets/game.png
android/assets/game2.png
did you merged branch with wrong atlas (from before river fix)?

Comment (HadeanLake) [reply core/src/com/unciv/models/ruleset/tech/Technology.kt]: Resolved conflict

File: core/src/com/unciv/logic/map/MapUnit.kt (hunk 2 above)
Comment (HadeanLake) [reply core/src/com/unciv/logic/map/MapUnit.kt]: I asked for screenshots, I will science out exact formula

Comment (HadeanLake) [reply core/src/com/unciv/logic/map/MapUnit.kt]: tooltip
It shows amount in the tooltip. Era and production in city does not matter (in civ 5 vanilla).
If tile is owned by worker's civ - 20 if next to city, 20 if one tile away, 15 if two tiles away, then 10, 5, 0
2/3 of that if tile is not owned - 13, 13, 10, 6, 3, 0

Notification looks like this:
NOTIFICATION

**Pull request merged**